readthedocs/rtd_tests/tests/test_backend.py

```diff
@@ -378,6 +389,252 @@ def test_fetch_clean_tags_and_branches(self, checkout_path):
+
+    Methods have been copied and adapted.
+    Once the feature ``GIT_CLONE_FETCH_CHECKOUT_PATTERN`` has become the default for all projects,
+    we can discard of TestGitBackend by moving the methods that aren't overwritten into this class
+    and renaming it.
+    """
+
+    def setUp(self):
+        super().setUp()
+        fixture.get(
+            Feature,
+            projects=[self.project],
+            feature_id=Feature.GIT_CLONE_FETCH_CHECKOUT_PATTERN,
+        )
+        self.assertTrue(
+            self.project.has_feature(Feature.GIT_CLONE_FETCH_CHECKOUT_PATTERN)
+        )
+
+    def test_check_for_submodules(self):
+        """
+        Test that we can get a branch called 'submodule' containing a valid submodule.
+        """
+        repo = self.project.vcs_repo(
+            environment=self.build_environment,
+            version_type=BRANCH,
+            version_identifier="submodule",
+        )
+
+        repo.update()
+        self.assertFalse(repo.are_submodules_available(self.dummy_conf))
+
+        # The submodule branch contains one submodule
+        repo.checkout("submodule")
+        self.assertTrue(repo.are_submodules_available(self.dummy_conf))
+
+    def test_check_invalid_submodule_urls(self):
+        """Test that a branch with an invalid submodule fails correctly."""
+        repo = self.project.vcs_repo(
+            environment=self.build_environment,
+            version_type=BRANCH,
+            version_identifier="invalidsubmodule",
+        )
+        repo.update()
+        repo.checkout("invalidsubmodule")
+        with self.assertRaises(RepositoryError) as e:
+            repo.update_submodules(self.dummy_conf)
+        # `invalid` is created in `make_test_git`
+        # it's a url in ssh form.
+        self.assertEqual(
+            str(e.exception),
+            RepositoryError.INVALID_SUBMODULES.format(["invalid"]),
+        )
+
+    def test_check_submodule_urls(self):
+        """Test that a valid submodule is found in the 'submodule' branch."""
+        repo = self.project.vcs_repo(
+            environment=self.build_environment,
+            version_type=BRANCH,
+            version_identifier="submodule",
+        )
+        repo.update()
+        repo.checkout("submodule")
+        valid, _ = repo.validate_submodules(self.dummy_conf)
+        self.assertTrue(valid)
+
+    @patch("readthedocs.vcs_support.backends.git.Backend.fetch_ng")
+    def test_git_update_with_external_version(self, fetch):
+        """Test that an external Version (PR) is correctly cloned and fetched."""
+        version = fixture.get(
+            Version,
+            project=self.project,
+            type=EXTERNAL,
+            active=True,
+        )
+        repo = self.project.vcs_repo(
+            verbose_name=version.verbose_name,
+            version_type=version.type,
+            version_identifier="84492ad53ff8aba83015123f4e68d5897a1fd5bc",
```

Comment:
> I understand the identifier for PRs is just the number of the PR instead a commit hash, right?

No, the identifier is the commit hash and `verbose_name` is the PR number :coffin: 

I put an overview in the `get_remote_fetch_refspec` docstring.

Unfortunately, this will continue to be annoying and counter-intuitive every time we revisit this code. But I think that it can be a task for refactoring once we have eliminated the other VCS backends from the risk calculation :)

Reply:
You are right. In any case, we should pass the `version.identifier` here. We should make those values `identifier` and `verbose_name` to be correct when the `Version` object is created.

Reply:
Agreed, updated :+1: